source/application/GUI/Core/DlgBase.cpp

```diff
@@ -0,0 +1,614 @@
+    {
+        return callback();
+    }
+private:
+    std::function<bool()> callback;
+};
+
+template<typename T>
+class CallbackLparam : public CallbackBase
+{
+public:
+    CallbackLparam(std::function<bool(T)> cb) :
+        callback(cb)
+    {
+    }
+    virtual bool Callback(LPARAM lparam, WPARAM wparam)
+    {
+        return callback(reinterpret_cast<T>(lparam));
+    }
+private:
+    std::function<bool(T)> callback;
+};
+
+template<typename T>
+class CallbackWparam : public CallbackBase
+{
+public:
+    CallbackWparam(std::function<bool(T)> cb) :
+        callback(cb)
+    {
+    }
+    virtual bool Callback(LPARAM lparam, WPARAM wparam)
+    {
+        return callback(reinterpret_cast<T>(wparam));
+    }
+private:
+    std::function<bool(T)> callback;
+};
+
+template<typename T, typename U>
+class CallbackBoth : public CallbackBase
+{
+public:
+    CallbackBoth(std::function<bool(T, U)> cb) :
+        callback(cb)
+    {
+    }
+    virtual bool Callback(LPARAM lparam, WPARAM wparam)
+    {
+        return callback(reinterpret_cast<T>(lparam), reinterpret_cast<U>(wparam));
+    }
+private:
+    std::function<bool(T, U)> callback;
+};
+
+class CallbackWmCommand : public CallbackBase
+{
+public:
+    CallbackWmCommand(std::function<bool(WORD)> cb) :
+        callback(cb)
+    {
+    }
+    virtual bool Callback(LPARAM lparam, WPARAM wparam)
+    {
+        return callback(HIWORD(wparam));
+    }
+private:
+    std::function<bool(WORD)> callback;
+};
+
+namespace
+{
+    static constexpr WCHAR typeface[] = L"MS Shell Dlg";
+
+#pragma pack(push, 1)
+    /*
+     * On MSDN these are declared as 1 struct, but, since the title-member is of variadic
+     * length it's declared as 2 structs here.
+     */
+    struct DLGTEMPLATEEX_1
+    {
+        WORD dlgVer;
+        WORD signature;
+        DWORD helpID;
+        DWORD exStyle;
+        DWORD style;
+        WORD cDlgItems;
+        short x;
+        short y;
+        short cx;
+        short cy;
+        WCHAR menu;
+        WCHAR windowClass;
+        WCHAR title;
+    };
+    struct DLGTEMPLATEEX_2
+    {
+        WORD      pointsize;
+        WORD      weight;
+        BYTE      italic;
+        BYTE      charset;
+        WCHAR     typeface[ARRAYSIZE(typeface)];
+    };
+    /*
+     * Hacky... But avoids weird magic
+     * -- Warepire
+     */
+    struct DLGITEMTEMPLATEEX_MINIMAL
+    {
+        DWORD reserved;
+        DWORD exStyle;
+        DWORD style;
+    };
+
+    struct MENUEX_TEMPLATE_HEADER
+    {
+        WORD wVersion;
+        WORD wOffset;
+        DWORD dwHelpId;
+    };
+    /*
+     * Hacky... But avoids weird magic
+     * -- Warepire
+     */
+    struct MENUEX_TEMPLATE_ITEM_MINIMAL
+    {
+        DWORD reserved[3];
+        WORD bResInfo;
+    };
+#pragma pack(pop)
+}
+
+std::map<HWND, DlgBase*> DlgBase::ms_hwnd_dlgbase_map;
+
+DWORD DlgBase::ms_ref_count = 0;
+
+namespace
+{
+    HINSTANCE gs_instance;

```

Comment:
Is this used anywhere?

Reply:
It will be once Open/Save dialogs and all that gets implemented.